File: [H]array/[H]array-sliding-window/3-longest-substring-without-repeating-characters.py
from collections import defaultdict
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        char_freq = defaultdict(int)
        res = 0
        left = 0
        for right in range(len(s)):
            char_freq[s[right]] += 1
            while char_freq[s[right]] > 1:
                char_freq[s[left]] -= 1
                left += 1
            res = max(res, right - left + 1)
        return res
        
# time O(n)
# space O(1), due to the hashmap could contain all unique chars
# using array and sliding window and hashmap

File: [H]array/[H]array-sliding-window/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> charToFreq = new HashMap<>();
        int res = 0;
        int left = 0;
        for (int right = 0; right < s.length(); right++) {
            Character c = s.charAt(right);
            charToFreq.compute(c, (k, v) -> v == null ? 1 : v + 1);
            while (charToFreq.get(c) > 1) {
                Character removeC = s.charAt(left);
                charToFreq.put(removeC, charToFreq.get(removeC) - 1);
                if (Objects.equals(charToFreq.get(removeC), 0)) {
                    charToFreq.remove(removeC);
                }
                left += 1;
            }
            res = Math.max(res, right - left + 1);
        }
        return res;
    }
}

// time O(n)
// space O(1), due to the hashmap could contain all unique chars
// using array and standard sliding window and hashmap